torch/autograd/gradcheck.py

```diff
@@ -180,68 +243,28 @@ def check_outputs_same_dtype_and_shape_in_neighborhood(output1, output2, idx, de
     # Computes the numerical jacobians wrt to a single input. Returns N jacobian
     # tensors, where N is the number of outputs. Input must require grad.
     assert input.requires_grad
-    jacobians = allocate_jacobians_with_outputs(outputs, input.numel(), input.dtype, input.device)
+    # We need a dictionary because for sparse inputs, d_idx aren't necessarily in order
+    jacobian_cols: Dict[int, List[Optional[torch.Tensor]]] = {}
 
     for x, idx, d_idx in iter_tensor(input):
-        # Computing the jacobian only works for pure real or pure imaginary delta
-        # for details on the algorithm used here, refer:
-        # Section 3.5.3 https://arxiv.org/pdf/1701.00392.pdf
-        # s = fn(z) where z = x for real valued input
-        # and z = x + yj for complex valued input
-        ds_dx_tup = compute_gradient(fn, inputs, input_idx, x, idx, eps, eps, input.layout)
-        if x.is_complex():  # C -> C, C -> R
-            ds_dy_tup = compute_gradient(fn, inputs, input_idx, x, idx, eps * 1j, eps, input.layout)
-
-            for ds_dx, ds_dy, d in zip(ds_dx_tup, ds_dy_tup, jacobians):
-                # conjugate wirtinger derivative
-                conj_w_d = 0.5 * (ds_dx + ds_dy * 1j)
-                # wirtinger derivative
-                w_d = 0.5 * (ds_dx - ds_dy * 1j)
-                d[d_idx] = grad_out.conjugate() * conj_w_d + grad_out * w_d.conj()
-        else:
-            for ds_dx, d in zip(ds_dx_tup, jacobians):
-                if ds_dx.is_complex():  # R -> C
-                    # w_d = conj_w_d = 0.5 * ds_dx
-                    # dL_dz_conj = 0.5 * [grad_out.conj() * ds_dx + grad_out * ds_dx.conj()]
-                    #            = 0.5 * [grad_out.conj() * ds_dx + (grad_out.conj() * ds_dx).conj()]
-                    #            = 0.5 * 2 * real(grad_out.conj() * ds_dx)
-                    #            = real(grad_out.conj() * ds_dx)
-                    d[d_idx] = torch.real(grad_out.conjugate() * ds_dx)
-                else:   # R -> R
-                    # skip if grad_out is complex but output is real
-                    if not isinstance(grad_out, complex):
-                        d[d_idx] = ds_dx * grad_out
-    return jacobians
+        def wrapped_fn():
+            inp = tuple(prepped_input(a, x if i == input_idx else None) if is_tensor_like(a) else a
+                        for i, a in enumerate(_as_tuple(inputs)))
+            return tuple(a.clone() for a in _as_tuple(fn(*inp)))
+
+        input_to_perturb = x[idx]
+        nbhd_checks_fn = functools.partial(check_outputs_same_dtype_and_shape_in_neighborhood, idx=idx, delta=delta)
+
+        def jvp_fn(delta):
+            return compute_numerical_gradient(wrapped_fn, input_to_perturb, delta, eps, nbhd_checks_fn)
+        jacobian_cols[d_idx] = []
+        compute_numerical_jacobian_cols(jacobian_cols[d_idx], delta, jvp_fn, x.is_complex(), grad_out)
```

Comment:
Mutation is difficult to reason about when there are a lot of variables flying around. Could we have something like the following?
```
jacobain_cols[d_idx] = compute_numerical_jacobian_cols(delta, jvp_fn, x.is_complex(), grad_out)
```

Reply:
Looks a lot cleaner with this change!